### DISCUSSION TITLE
Include autodetected dest_filesystem in a getter

### DISCUSSION THREAD

**Discussion by decoupca** [napalm-automation/napalm#1996] [Ideas]
For IOS, NAPALM requires the following in the running-config in order to merge or replace a config:

```
archive
  path flash:
```

Of course the path directive might be different, e.g. bootflash: depending on the specific device. I want to template out the right config ahead of time to enable config replace.

I see I can get the value with `device._discover_file_system()`. But (AFAICT) no getters return this value so I can't use ansible to discover it without a custom module. I could write a custom module, but wondering if others might also find it useful if this value were returned as an key in `get_facts()`, or another getter? Something like this:

```
{
    'uptime': 1651140.0,
    'vendor': 'Cisco',
    'os_version': 'Catalyst L3 Switch Software (CAT3K_CAA-UNIVERSALK9-M), Version 16.12.5b, RELEASE SOFTWARE (fc3)',
    'autodetected_dest_filesystem': 'flash:',
    ...
}

**Comment by ktbyers** [comment]
The main problem I see is that NAPALM is going to require all the core platforms to return the same data structure for `get_facts()`. 

So then we would have a bit of an issue with what to do about the other platforms (Junos, Arista, IOS-XR, NX-OS) and does what we are doing for IOS/IOS-XE make sense for these other platforms or not (as far as adding a new key for the `filesystem` and what value to populate this with).

You could probably do this with ntc-ansible and ntc_show_command module and just execute a `dir` (or alternatively just straight `ios_command` and once again execute a `dir`). 

This in NAPALM:

```
device._discover_file_system()
```

Calls this in Netmiko:

```
_autodetect_fs
```

Which basically just parses the `dir` output.

Let me know if that workaround will work for you (without needing to create an Ansible module).

**Reply by decoupca** [reply]
Agreed, the filesystem name by itself wouldn't make much sense for other platforms.

> Which basically just parses the dir output.

Thanks for the tip -- I suppose I could do the same with the napalm_cli module.

What do you think of a new getter, `get_filesystems()` or something similar? Could include filesystem names, total space, free space, and possibly a listing of files in the primary fs. Would include data from `show file systems` and `dir` in IOS, `show system storage` and `file list` in JUNOS, etc.

**Reply by ktbyers** [reply]
Sounds good to me. I would create a GitHub issue and propose the name of the `getter` and a proposed data structure for the results. We can then sanity check if that looks like it will work across the five core platforms.
